tests/functional/TestSmokeTest/tests/timeout_test.go

```diff
@@ -0,0 +1,126 @@
+// +build functional
+
+//
+// Copyright (c) 2015 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+package functional
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/heketi/heketi/pkg/glusterfs/api"
+	"github.com/heketi/heketi/pkg/remoteexec/ssh"
+	"github.com/heketi/heketi/pkg/testutils"
+	"github.com/heketi/tests"
+)
+
+var (
+	nodes         = 4
+	SSHPort       = "22"
+	preScriptFile = "/var/lib/glusterd/hooks/1/start/pre/S99timeout.sh"
+)
+
+func getNodes() []string {
+	nodelist := make([]string, nodes)
+
+	for index, ip := 0, 100; index < nodes; index, ip = index+1, ip+1 {
+		nodelist[index] = "192.168.10." + fmt.Sprintf("%v", ip)
+	}
+
+	return nodelist
+}
+
+func stopSleep(node string, exec *ssh.SshExec) error {
+
+	cmd := []string{" pkill -9 sleep || :"}
+	_, err := exec.ConnectAndExec(node+":"+SSHPort, cmd, 10, true)
+	return err
+}
+func addPostScript(timeOut int, node string, exec *ssh.SshExec) error {
+
+	//create and change file permission pre-script file
+	createFile := fmt.Sprintf("install -m 777 /dev/null %s", preScriptFile)
+
+	//write script which sleeps for minutes
+	dumpData := fmt.Sprintf("echo \"#! /bin/bash\n sleep %dm \">%s", timeOut, preScriptFile)
+
+	cmd := []string{createFile, dumpData}
+	_, err := exec.ConnectAndExec(node+":"+SSHPort, cmd, 10, true)
+	return err
+}
+
+func deletePreScript(node string, exec *ssh.SshExec) error {
+
+	//delete pre script file
+	removeFile := fmt.Sprintf("rm -rf %s", preScriptFile)
+	cmd := []string{removeFile}
+	_, err := exec.ConnectAndExec(node+":"+SSHPort, cmd, 10, true)
+	return err
+}
+
+func removeTimeout(sleep uint64, exec *ssh.SshExec, t *testing.T) {
+	time.Sleep(time.Duration(sleep) * time.Minute)
+	for _, node := range getNodes() {
+		err := deletePreScript(node, exec)
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+		err = stopSleep(node, exec)
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+	}
+
+}
+func TestGlusterTimeout(t *testing.T) {
+
+	// Setup the VM storage topology
+	teardownCluster(t)
+	setupCluster(t, 4, 8)
+	defer teardownCluster(t)
+
+	na := testutils.RequireNodeAccess(t)
+	//add postscript in all gluster nodes
+	exec := na.Use(logger)
+	for _, node := range getNodes() {
+		err := addPostScript(6, node, exec)
+		tests.Assert(t, err == nil, "expected err == nil, got:", err)
+
+	}
+
+	//removeTimeout deletes the postscript file and stops the sleep process
+	//which was started as part of pre script
+	//sleep for 3 minutes to validate we are not getting any timeout error
+	//from gluster cli
+	go removeTimeout(3, exec, t)
```

Comment:
I don't understand the use of a goroutine here. Was this supposed to be a defer?

Reply:
to block the glusterd processing we are making use of the prescript in glusterd.

this is the goroutine which will delete the prescript script and kill the sleep process which will be running as part of the prescript by glusterd.
once this script is deleted and the sleep process is killed, volume start will continue.

Reply:
I find it a bit confusing, and I would think that future readers would find it confusing as well. Let's please rework the general approach. As I mentioned previously, we should:
* Install the hook script at set up time via ansbile
* Have the hook script do nothing in the normal case, sleep if a condition file exists
* Have the hook script write the pid of the process that needs killing into a pidfile
* In your test all you need to do is touch the condition file to initiate the sleep
* In your test you read/use the pid file to kill the sleep
* Avoid sleeping in threads in the tests and trigger the events when you need them

Reply:
>Install the hook script at set up a time via ansible

@phlogistonjohn hook script needs to be created and deleted multiple times. if we install scripts during setup time.
but still we need to delete and create the same file, even i need to update the sleep time in hook script for few tests
let me know your thoughts on this.

Reply:
You could install the script without executable permissions, and just toggle those when the script needs to be enabled/disabled.

Reply:
I want the mechanism to be in place throughout, but the configuration should change between tests. That's why I am suggesting to install a hook script that always runs but does nothing unless a "configuration file" is in place. This configuration file could be as simple as a single number that is your sleep time. Let's suppose that we decide the config file lives at /var/tmp/hook-test-delay, if the file is not present the hook script does nothing. If the file contains "5" then the script sleeps for 5 seconds before exiting. Your test is now only responsible for setting up and removing this config file rather than understanding how hook scripts in gluster are installed.

An advantage of doing this is that you make the script re-usable for many more test scenarios and there's less debugging effort needed if things don't work because you only need to test the script part in isolation from the heketi test code. If you combine the two you are having to debug the way the heketi test code works in combination with the script every time.
In addition, if gluster were to change the way hook scripts worked the heketi code would be isolated from that.

Reply:
Madhu and I discussed this yesterday a little as well. A script that gets the delay from a configuration file is one option, setting (and parsing) a custom `user.start-delay` volume option would be an other.

Reply:
@nixpanic @phlogistonjohn I have updated the code as per John approach please check.